applications/procedural_narrator/test_procedural_narrator.py

```
@@ -0,0 +1,251 @@
+
+    assert dim_specs[2].color == bright_specs[2].color == narrator_module.TEXT_COLOR
+    assert dim_specs[7].color[:3] == bright_specs[7].color[:3]
+    assert dim_specs[7].color[3] == pytest.approx(0.35)
+    assert bright_specs[7].color[3] == pytest.approx(0.75)
+    assert working_pulse_alpha(0.75) == pytest.approx(narrator_module.WORKING_PULSE_MAX_ALPHA)
+    assert working_pulse_alpha(2.25) == pytest.approx(narrator_module.WORKING_PULSE_MIN_ALPHA)
+
+
+def test_application_rejects_unknown_source():
+    with pytest.raises(ValueError, match="source"):
+        ProceduralNarratorApp("/tmp/data", source="file")
```

Ruff (0.15.21)

[error] Probable insecure usage of temporary file or directory: "/tmp/data"

(S108)